### PR TITLE
[BEAM-2939] Support SDF expansion in the Flink runner.

### PR DESCRIPTION
**Pull request by robertwb**
SDFs now expand to Split + Reshard + Read. Informing the SDF about the desired number of splits is still TBD. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py [hunk 1]
```diff
@@ -788,14 +799,34 @@ def make_stage(base_stage, transform_id, extra_must_follow=()):
             inputs=dict(transform.inputs, **{main_input_tag: paired_pcoll_id}),
             outputs={'out': split_pcoll_id})
 
+        if common_urns.composites.RESHUFFLE.urn in context.known_runner_urns:
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: Is the plan to eventually retrieve the known URNs dynamically?

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py [hunk 2]
```diff
@@ -788,14 +799,34 @@ def make_stage(base_stage, transform_id, extra_must_follow=()):
...
+              transform,
+              unique_name=transform.unique_name + '/Reshuffle',
+              urn=common_urns.composites.RESHUFFLE.urn,
+              payload=b'',
+              inputs=dict(transform.inputs, **{main_input_tag: split_pcoll_id}),
+              outputs={'out': reshuffle_pcoll_id})
+          yield make_stage(stage, reshuffle_transform_id)
+        else:
+          reshuffle_pcoll_id = split_pcoll_id
+          reshuffle_transform_id = None
```
Comment (mxm) [inline sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: Isn't Reshuffle necessary to balance split work evenly?

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: Either that or make RESHUFFLE a required primitive. (Or inject a full composite reshuffle here, but that has a pile of ufuncs in it.)

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py (hunk 2 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: Some runners have the notion of fusion breaks (see the mustfollow below) that can do this. The direct/in-process runners don't care.

**Comment by robertwb** [comment]
Good questions, Luke.

I don't think we have the data for (1) yet. In particular, I'm not aware of a way to query SDFs (in the SDKs) as to whether they produce unbounded output from bounded input. (2) is already done in Python. For (3) we'll need to add this to the java runner shared code as well, and we'll definitely need (4/5) for streaming.

I was planning on getting batch working first, then tackling streaming. This will allow us to at least test TFX pipelines.

**Comment by lukecwik** [comment]
For 1) we already know whether an SDF is bounded or not in the model pipeline based upon whether any of its output [PCollections are bounded or not](https://github.com/apache/beam/blob/2b111c46c5a928e1d47e2b51e26d36ca206f0089/model/pipeline/src/main/proto/beam_runner_api.proto#L348) and in the Java SDK we explicitly require uses to annotate SDFs with whether they are unbounded or not using [@UnboundedPerElement](https://github.com/apache/beam/blob/2b111c46c5a928e1d47e2b51e26d36ca206f0089/sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java#L600). Should the Python SDK also add an annotation similar to the Java SDK?

**Pull request merged**